### ISSUE TITLE
Hardcoded 'xterm'

### ISSUE THREAD

**Issue by gvanuxem** [fricas/fricas#119]
From what I see 'xterm' is hardcoded in several places in the 'src' directory files:

So if I issue ')edit /path/to/my/file:
*********************************************************
Warning: environment variable FRICASEDITOR not set.
Launching 'more' in an 'xterm'.
Press space to continue, 'q' to quit.
*********************************************************
So 'xterm' is needed. If I run graphics examples via HyperDoc I also need 'xterm'. If I install 'xterm', with X started I have graphic examples produced from HyperDoc of course.

With or without xterm (and X) the 'vi' command with ')edit /home/greg/.bashrc' (via $FRICASEDITOR env var set and not, as the ')help edit' suggests , $EDITOR var)  produces nothing really usable.

So I would first suggest to add 'xterm' in prerequisites in the INSTALL file with X libraries.

But for $FRICASEDITOR, with or without X, 'xterm' is needed. To read or edit a file it could be handy to open it directly in the FriCAS REPL. Suggestions are welcome.

More things can be done I think.

**Comment by hemmecke** [comment]
Agreed. To hardcode "xterm" is probably a bad idea given that there is not just *nix systems.
However, honestly, I rather think that ")edit" should be removed completely. Nowadays you can access any file with any editor and you can do this while FriCAS is running. There is simply no need anymore to start an editor from within FriCAS. Am I wrong?

**Comment by gvanuxem** [comment]
Yes, I do agree this is not a "must have" feature. I suspect its main use
is to edit just compiled files with just an empty parameter ')edit'. To fix
a typo for example. It's documented in the help file, but if it works with
'more' it does not edit the `` last file'' with 'vi' for me.

Le mar. 28 févr. 2023 à 18:34, Ralf Hemmecke ***@***.***> a
écrit :

**Comment by hemmecke** [comment]
BTW, look into i-syscmd.boot. ")edit" calls editFile from compat.boot
```
editFile file ==
  MEMQ(INTERN('"WIN32",FIND_-PACKAGE("KEYWORD")),_*FEATURES_*) =>
      OBEY STRCONC('"notepad ", file)
  OBEY STRCONC('"$FRICAS/lib/SPADEDIT ", file)
```
If you look into the script SPADEDIT, then you find that "xterm" is only used if you haven't set your FRICASEDITOR shell variable.
Unfortunately, I cannot give you a good value for FRICASEDITOR that works when running FriCAS inside WSL2 under Windows. So xterm is not totally hardcoded.
I guess, that it might make sense to change SPADEDIT so that the WSL2 case is nicely covered by default. (I have no clue, how.)

**Comment by gvanuxem** [comment]
On WSL 2, it's a real Linux system. Yes. With an extended 4 filesystem.

$uname -a
Linux ellipse 5.15.90.1-microsoft-standard-WSL2 #1 SMP Fri Jan 27 02:56:13
UTC 2023 x86_64 GNU/Linux

Applications have no cues to know they are on a Linux system virtualized
via the processor capability. It's not on an emulated system. Maybe I said
that but no, it's really Linux. The only difference I know of is that
'init' and its different levels have no meanings, you have to start
yourself the necessary services and X has no real use since you're on the
'explorer.exe' Window Manager (unless via 'xrdp').

And let me add that, now, with support of GUI app I expressed yesterday I
could use gvim or gedit directly in Windows, but started in Linux. In WSL2
'C:\' is in '/mnt/c/'.

SBCL:
(:X86-64 :GENCGC :64-BIT :ANSI-CL :COMMON-LISP :ELF :IEEE-FLOATING-POINT
:LINUX
 :LITTLE-ENDIAN :PACKAGE-LOCAL-NICKNAMES :SB-CORE-COMPRESSION :SB-LDB
 :SB-PACKAGE-LOCKS :SB-THREAD :SB-UNICODE :SBCL :UNIX)

CCL:
(:PRIMARY-CLASSES :COMMON-LISP :OPENMCL :CCL :CCL-1.2 :CCL-1.3 :CCL-1.4
:CCL-1.5 :CCL-1.6 :CCL-1.7 :CCL-1.8 :CCL-1.9 :CCL-1.10 :CCL-1.11 :CCL-1.12
:CLOZURE :CLOZURE-COMMON-LISP :ANSI-CL :UNIX :OPENMCL-UNICODE-STRINGS :IPV6
:OPENMCL-NATIVE-THREADS :OPENMCL-PARTIAL-MOP :MCL-COMMON-MOP-SUBSET
:OPENMCL-MOP-2 :OPENMCL-PRIVATE-HASH-TABLES
:STATIC-CONSES-SHOULD-WORK-WITH-EGC-IN-CCL :PACKAGE-LOCAL-NICKNAMES :X86-64
:X86_64 :X86-TARGET :X86-HOST :X8664-TARGET :X8664-HOST :LINUX-HOST
:LINUX-TARGET :LINUXX86-TARGET :LINUXX8664-TARGET :LINUXX8664-HOST
:64-BIT-TARGET :64-BIT-HOST :LINUX :LITTLE-ENDIAN-TARGET
:LITTLE-ENDIAN-HOST)

So to respond, if FRICASEDITOR is set to 'vi', as said earlier, it is not
usable apparently and, also, the help file should be modified. Cf. unused
$EDITOR variable.


Le mar. 28 févr. 2023, 21:42, Ralf Hemmecke ***@***.***> a
écrit :

**Comment by hebisch** [comment]
On Tue, Feb 28, 2023 at 09:34:59AM -0800, Ralf Hemmecke wrote:
> Agreed. To hardcode "xterm" is probably a bad idea given that there is not just *nix systems.
> However, honestly, I rather think that ")edit" should be removed completely. Nowadays you can access any file with any editor and you can do this while FriCAS is running. There is simply no need anymore to start an editor from within FriCAS. Am I wrong?

Main motivation to keep this machinery is to view system sources.
For user it may be tricky to find out which file contains given
constructor.  Hyperdoc knows how to find source file and launching
file viewer is IMO nice feature.

I am not sure if anybody cares about editing, but the same machinery
is also used for viewing.

Another thing is that in general I am trying to preserve code
that is potentially useful for future developement.  Edit
part in principle could be replaced by something much more
advanced.  Of course, in 15 years there was no significant
developement here, so chance for improvement are very low.
OTOH removing code takes effort.  I remove code that is
problematic (and a lot of code was removed), but keep
working unproblematic and not duplicated code (for me
duplicated code is problematic).

To put is differently, removing code without removing
functionality tends to improve quality of our code base
and happens a lot.  Eager removal of functionality risk
in long term we will end up with no functionality, so
I am rather conservative here.

-- 
                              Waldek Hebisch

**Comment by hebisch** [comment]
On Wed, Mar 01, 2023 at 04:07:11AM -0800, Greg wrote:
> On WSL 2, it's a real Linux system. Yes. With an extended 4 filesystem.
> 
> $uname -a
> Linux ellipse 5.15.90.1-microsoft-standard-WSL2 #1 SMP Fri Jan 27 02:56:13
> UTC 2023 x86_64 GNU/Linux
> 
> Applications have no cues to know they are on a Linux system virtualized
> via the processor capability. It's not on an emulated system. Maybe I said
> that but no, it's really Linux. The only difference I know of is that
> 'init' and its different levels have no meanings, you have to start
> yourself the necessary services and X has no real use since you're on the
> 'explorer.exe' Window Manager (unless via 'xrdp').

I am not sure what window manager have to do with "use of X".  AFAIK
you can run X server on Windows (just Windows, without need for WSL).
If WSL is real Linux, then it can make network connections and
connect to X under Windows.  So any program using X should work
in this setup.  Of course, using network connctions you need to
fight with Windows firewall which may be too hard average
Windows user.

And in current times Linux have a lot of graphic programs.  Does
WSL support running graphic programs?  If yes, then question is
how?

-- 
                              Waldek Hebisch

**Comment by hebisch** [comment]
```
Few more comments:
On Tue, Feb 28, 2023 at 09:05:46AM -0800, Greg wrote:
> >From what I see 'xterm' is hardcoded in several places in the 'src' directory files:
> 
> So if I issue ')edit /path/to/my/file:
> ==================================
> Warning: environment variable FRICASEDITOR not set.
> Launching 'more' in an 'xterm'.

As wrote earlier, 'more' in 'xterm' is a default, you can choose
different editor via environment variable.  Maybe the message
above shoud be:

Launching 'more' in an 'xterm'.  You can specify different
editor by setting FRICASEDITOR environment variable.

> So 'xterm' is needed. If I run graphics examples via HyperDoc I also
> need 'xterm'. If I install 'xterm', with X started I have graphic
> examples produced from HyperDoc of course.

More precisely, FriCAS uses extra terminal when it _executes_ example
code from HyperDoc: does not mater if graphics or not, what counts
is fact that example is executed as opposed to using precomputed
result (precomputed results, including graphics work without extra
termial).

We could make this configurable, but what are reasonable choices?
We need ability to pass command line to execute inside the terminal
and ability to set window title is desirable.  We could easily
check is say "FRICAS_TERM" environment variable is set and use it
as command line string to which we concatenate command to start in
the terminal.  But it is not clear if such simple concatenation
is convenient for other terminal emulators: it is not convenient
for 'xterm', one would have to use shell functions to get
desired functionality.

> So I would first suggest to add 'xterm' in prerequisites in the
> INSTALL file with X libraries.

In past 35 years 'xterm' was considerd standard (mandatory) part
of X11.  This was exactly due to use cases like our: program
needs a terminal emulator and wants to use one already installed
in the system.  System may have many different terminal emulators
but 'xterm' was the only one warranted to exist.

In INSTALL we probably should say that user needs _standard_ X windows
including 'xterm'.

-- 
                              Waldek Hebisch

**Comment by gvanuxem** [comment]
First, thanks for your response.

Le sam. 17 juin 2023 à 01:04, hebisch ***@***.***> a écrit :

> ```
> Few more comments:
> On Tue, Feb 28, 2023 at 09:05:46AM -0800, Greg wrote:
> > >From what I see 'xterm' is hardcoded in several places in the 'src'
> directory files:
> >
> > So if I issue ')edit /path/to/my/file:
> > ==================================
> > Warning: environment variable FRICASEDITOR not set.
> > Launching 'more' in an 'xterm'.
>
> As wrote earlier, 'more' in 'xterm' is a default, you can choose
> different editor via environment variable. Maybe the message
> above shoud be:
>
Launching 'more' in an 'xterm'. You can specify different
> editor by setting FRICASEDITOR environment variable.
>

Personally I would prefer this message.

And, as an opinion, I would preferably use the $EDITOR (and eventually also
$VISUAL) well known environment variables.


>
> > So 'xterm' is needed. If I run graphics examples via HyperDoc I also
> > need 'xterm'. If I install 'xterm', with X started I have graphic
> > examples produced from HyperDoc of course.
>

Yes,  I know.

>
> More precisely, FriCAS uses extra terminal when it _executes_ example
> code from HyperDoc: does not mater if graphics or not, what counts
> is fact that example is executed as opposed to using precomputed
> result (precomputed results, including graphics work without extra
> termial).
>
> We could make this configurable, but what are reasonable choices?
> We need ability to pass command line to execute inside the terminal
> and ability to set window title is desirable. We could easily
> check is say "FRICAS_TERM" environment variable is set and use it
> as command line string to which we concatenate command to start in
> the terminal. But it is not clear if such simple concatenation
> is convenient for other terminal emulators: it is not convenient
> for 'xterm', one would have to use shell functions to get
> desired functionality.
>

Yes, I think FriCAS is right, 'xterm' in fact is _the_ good choice. My
issue comes from a personal mistake. I likely removed some X11 apps like
xterm. I checked via 'apt-cache rdepends xterm' and yes, installing xorg on
my Debian-like system installs xterm. As needed.
I checked other terminal emulator and they do not accept the same arguments
so I think it would be a mess to handle other terminals correctly. So, it's
my fault.

**Comment by gvanuxem** [comment]
But, yes, as said by Ralf Hemmecke, on other systems like MinGW/MSYS where X11 (or the like) are not available xterm has no sense.

**Comment by gvanuxem** [comment]
See also: https://groups.google.com/g/fricas-devel/c/npxyv27DPtU

**Comment by oldk1331** [comment]
Resolved by 655c1f828b49fa7db594c15da7befd6a7cc79839 .
